state/unit.go

```diff
@@ -891,27 +965,52 @@ func (u *Unit) EnsureDead() (err error) {
 // the application is Dying and no other references to it exist. It will fail if
 // the unit is not Dead.
 func (u *Unit) Remove() (err error) {
+	return u.RemoveWithForce(false)
+}
+
+// ForceRemove removes the unit from state similar to the unit.Remove() but
+// it ignores errors.
+func (u *Unit) RemoveWithForce(force bool) error {
+	// TODO (anastasiamac 2019-04-2) First return here is operational errors.
+	// We might want to consider to pass them up to notify users of non-fatal
+	// errors we have encountered.
+	_, err := u.internalRemove(force)
+	return err
+}
+
+func (u *Unit) internalRemove(force bool) (errs []error, err error) {
 	defer errors.DeferredAnnotatef(&err, "cannot remove unit %q", u)
 	if u.doc.Life != Dead {
-		return errors.New("unit is not dead")
+		return errs, errors.New("unit is not dead")
 	}
-
 	// Now the unit is Dead, we can be sure that it's impossible for it to
 	// enter relation scopes (once it's Dying, we can be sure of this; but
 	// EnsureDead does not require that it already be Dying, so this is the
 	// only point at which we can safely backstop lp:1233457 and mitigate
 	// the impact of unit agent bugs that leave relation scopes occupied).
 	relations, err := applicationRelations(u.st, u.doc.Application)
 	if err != nil {
-		return err
-	}
-	for _, rel := range relations {
-		ru, err := rel.Unit(u)
-		if err != nil {
-			return err
+		if !force {
+			return errs, err
 		}
-		if err := ru.LeaveScope(); err != nil {
-			return err
+		errs = append(errs, err)
+	} else {
+		var failRelations error
+		for _, rel := range relations {
+			ru, err := rel.Unit(u)
+			if err != nil {
+				errs = append(errs, err)
+				failRelations = err
+				continue
+			}
+			if err := ru.LeaveScope(); err != nil {
+				errs = append(errs, err)
+				failRelations = err
+				continue
+			}
+		}
+		if !force && failRelations != nil {
```

Comment:
Is this supposed to be inside the loop? Or are you intending to call LeaveScope on all of the relation units even in the non-force case?

Reply:
I was going to keep going until I was done with all relations.. Why? Just because one failed, should not we try the rest?
If we are in non-force case and encountered any error, we'll still fall over with latest error after going through all of them.

Reply:
After an offline discussion, we'll leave as is :) Thank you for your time and for pondering over this \o/